### PR TITLE
Use chart labels to filter alerts

### PR DESCRIPTION
**Pull request by MrZammler**
##### Summary

This PR adds a new configuration key to alerts, called `chart labels`. This key can be used to specify chart labels that should match (based on simple patterns) in order to link or not an alert. It can be used to provide the same functionality as the `families` key (assuming of course the same values exist as chart labels as family values).

For example, the families key in `disks.conf`:

`families: !/dev !/dev/* !/run !/run/* *` can be replaced with:

`chart labels: !mount_point=/dev !mount_point=/dev/* !mount_point=/run !mount_point=/run/* mount_point=*`

It is part of this initiative: https://github.com/netdata/product/issues/3118

##### Test Plan

Test by replacing `families` (we use them in disks and net confs mostly) with the appropriate chart labels (disks can be `mount_point`, network devices can be `device` or even `interface_type`.

In general if we don't change our stock alerts, code should be safe to merge.

##### Additional Information

<details> <summary>For users: How does this change affect me?</summary>
  
</details>

### CODE REVIEW

**Comment by ilyam8** [comment]
> chart labels: !mount_point=/dev !mount_point=/dev/* !mount_point=/run !mount_point=/run/* mount_point=*

Hey, @MrZammler. Is it the same syntax that we use in "host labels"? The one you suggested looks a bit strange to me. What do you think about this?

```
chart labels: mount_point = !/dev !/dev/* !/run !/run/* *
```

- If there are several labels - just use several `chart labels` lines.
- Also, probably just `labels` is better.

**Comment by ilyam8** [comment]
@MrZammler Additionally, I think that if we add `labels`/`chart labels` filter we can remove not only `families` but `charts` too.

**Comment by thiagoftsm** [comment]
> @MrZammler Additionally, I think that if we add `labels`/`chart labels` filter we can remove not only `families` but `charts` too.

@ilyam8 I think this is a compatibility break that we need to announce before to remove.

**Review by cakrit — APPROVED** [review]
Approved for the readme

**Comment by MrZammler** [comment]
> > chart labels: !mount_point=/dev !mount_point=/dev/* !mount_point=/run !mount_point=/run/* mount_point=*
> 
> Hey, @MrZammler. Is it the same syntax that we use in "[host labels](https://learn.netdata.cloud/docs/alerts-and-notifications/configure-alerts#alarm-line-host-labels)"? The one you suggested looks a bit strange to me. What do you think about this?
> 
> ```yaml
> chart labels: mount_point = !/dev !/dev/* !/run !/run/* *
> ```
> 
>     * If there are several labels - just use several `chart labels` lines.
> 
>     * Also, probably just `labels` is better.

Hi @ilyam8 ! Yes, this is the same as we use in host labels. The way the code is now it only works as the example I gave. I agree what you wrote makes sense, but would need more changes, multiple chart labels fields if we wish to check on more labels, etc. I'll see if there is a nice way to do it.

**Comment by ilyam8** [comment]
@MrZammler please do, I think that the current implementation is not good. We can discuss it.

**Comment by ilyam8** [comment]
@MrZammler an alternative could be

```yaml
chart labels: mount_point = "!/dev !/dev/* !/run !/run/* *" filesystem = "!tmpfs *"
```

**Comment by ilyam8** [comment]
@MrZammler I tried `chart labels` and it works. We need to update the docs:
 - syntax
  ```
  chart labels: cgroup_name = !nginx1 !nginx2 *
  ```
  =>
  ```
  chart labels: cgroup_name = !nginx1 !nginx2 *
  ```

- mention that lack of a label means a negative match.
- mention that logic is OR when using multiple labels. E.g.
  ```
  chart labels: cgroup_name = nginx1 image = redis
  ```
  is cgroup_name is nginx1 OR image is redis.

**Comment by MrZammler** [comment]
> mention that lack of a label means a negative match.

Thanks Ilya, I've updated the docs. Can you please though explain the above? How could we write it in the doc?

**Comment by ilyam8** [comment]
> Can you please though explain the above? 

```yaml
chart labels: kkk = !foo *
```

A chart has no `kkk` (label) => a negative match => the chart is filtered.

File: health/health_config.c [hunk 1]
```diff
@@ -937,6 +984,27 @@ static int health_readfile(const char *filename, void *data) {
...
+                              line, filename, rrdcalc_name(rc), key, value, value);
+
+                    string_freez(rc->chart_labels);
+                    simple_pattern_free(rc->chart_labels_pattern);
+                }
+
+                {
+                    char *tmp = simple_pattern_trim_around_equal(value);
+                    char *tmp_2 = health_config_add_key_to_values(tmp);
+                    rt->chart_labels = string_strdupz(tmp_2);
```
Comment (thiagoftsm) [inline health/health_config.c:1001]: Hello @MrZammler ,

This line has the following issue:

```sh

Thread 32 "HEALTH" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe54df6c0 (LWP 30121)]
0x00000000004a6683 in health_readfile (filename=filename@entry=0x132e3b0 "/etc/netdata/health.d/my_disk.conf", data=data@entry=0xcb9f40) at health/health_config.c:1001
1001                        rt->chart_labels = string_strdupz(tmp_2);
(gdb) bt
#0  0x00000000004a6683 in health_readfile (filename=filename@entry=0x132e3b0 "/etc/netdata/health.d/my_disk.conf", data=data@entry=0xcb9f40) at health/health_config.c:1001
#1  0x0000000000452dc5 in recursive_config_double_dir_load (user_path=user_path@entry=0x139a560 "/etc/netdata/health.d", stock_path=stock_path@entry=0x130fd40 "/usr/lib/netdata/conf.d/health.d", subpath=subpath@entry=0x0, 
    callback=callback@entry=0x4a4663 <health_readfile>, data=data@entry=0xcb9f40, depth=depth@entry=0) at libnetdata/libnetdata.c:1552
#2  0x00000000004a7fa7 in health_readdir (host=host@entry=0xcb9f40, user_path=0x139a560 "/etc/netdata/health.d", stock_path=stock_path@entry=0x130fd40 "/usr/lib/netdata/conf.d/health.d", subpath=subpath@entry=0x0)
    at health/health_config.c:1338
#3  0x00000000004a1074 in initialize_health (host=host@entry=0xcb9f40) at health/health.c:804
#4  0x00000000004a1478 in health_main (ptr=0xb8cbe0) at health/health.c:1022
#5  0x0000000000463694 in netdata_thread_init (ptr=<optimized out>) at libnetdata/threads/threads.c:262
#6  0x00007ffff6badcaa in start_thread () from /lib64/libc.so.6
#7  0x00007ffff6c37d9c in clone3 () from /lib64/libc.so.6
```

I discovered it after to modify a template alert:

```sh
template: disk_example
      on: disk.busy
  lookup: sum -1m unaligned absolute of busy
   units: seconds
chart labels: device=sda
#  charts: !*sdc* *
   every: 1m
    warn: $this > 4000
    info: testing with disks
      to: sysadmin
```

to a simple alarm:

```sh
alarm: disk_example
      on: disk_busy.sda
  lookup: sum -1m unaligned absolute of busy
   units: seconds
chart labels: device=sda
#  charts: !*sdc* *
   every: 1m
    warn: $this > 4000
    info: testing with disks
      to: sysadmin
```

Please, take a look when you have a time.

Comment (MrZammler) [reply health/health_config.c:1001]: Thanks @thiagoftsm ! Looking!

Comment (MrZammler) [reply health/health_config.c]: "I will not copy paste again", "I will not copy paste again" .... Thanks Thiago!

**Review by thiagoftsm — APPROVED** [review]
After last commit I test different alerts(alarm, template, foreach) and I observed that everything is working as expected, LGTM!

**Review by ilyam8 — APPROVED** [review]
lgtm, but I am in two minds about OR logic. Why do you think we need OR and not AND?

**Comment by MrZammler** [comment]
@ilyam8 On the same boat.... we'll revisit later. Thanks.

**Pull request merged**